twisted/internet/iocpreactor/abstract.py

```diff
@@ -236,8 +237,9 @@ def doWrite(self):
         if len(self.dataBuffer) - self.offset < self.SEND_LIMIT:
             # If there is currently less than SEND_LIMIT bytes left to send
             # in the string, extend it with the array data.
-            self.dataBuffer = (buffer(self.dataBuffer, self.offset) +
-                               "".join(self._tempDataBuffer))
+            writeView = memoryview(self.dataBuffer)
+            self.dataBuffer = (writeView[self.offset:].tobytes() +
+                               b"".join(self._tempDataBuffer))
```

Comment:
of course, you can't do `memoryview + bytes`, so your choices are either to drop `memoryview` entirely or keep `tobytes()`.

i think you should drop the `memoryview` entirely.  as i said in a previous comment, using it and then calling `tobytes()` copies the backing array, apparently negating speed and memory savings.  it might seem that slicing into a `memoryview` with a large `self.offset` will only copy the last part of the backing array, but this is actually how `bytes` and `str` slicing already works in cpython.  because strings are immutable, slices can be shared.

tl;dr: just do:

```python
self.dataBuffer = self.dataBuffer[self.offset:] + b"".join(self._tempDataBuffer)
```